titan/model.py

```diff
@@ -358,48 +362,112 @@ def agents_interact(self, rel: "ag.Relationship"):
+                            agent.hiv.active,
+                            agent.hiv.aids,
+                            agent.drug_type,
+                            agent.sex_type,
+                            agent.haart.adherent,
+                            agent.race,
+                            agent.location,
+                            self.params.model.time.steps_per_year,
+                            strategy.exit_class,
+                        )
+                        * self.calibration.mortality
+                    )
+
+                    if self.run_random.random() < p:
+                        # agent dies
+                        self.exits[strategy.exit_class].append(agent)
+                        break
+                elif exit.exit_type == "drop_out":
+                    p = (
+                        agent.location.params.demographics[agent.race]
+                        .sex_type[agent.sex_type]
+                        .drug_type[agent.drug_type]
+                        .exit[strategy.exit_class]
+                        .prob
+                    )
+                    if self.run_random.random() < p:
+                        # agent leaves study pop
+                        self.exits[strategy.exit_class].append(agent)
+                        break
+        for l in self.exits.values():
+            for agent in l:
+                self.remove_agent(agent)
+
+    def enter(self):
+        for strategy in self.params.enter_exit.values():
+            entrance = self.params.classes.enter[strategy.entry_class]
+            if entrance.enter_type == "new_agent":
+                # determine new agent locations and characteristics
+                if self.params.classes.exit[strategy.exit_class].exit_type == "none":
+                    # Adding new agents without removing any
+                    num_new_agents = len(self.pop.all_agents.members) * entrance.prob
+                else:
+                    # number of new agents given removed agents
+                    num_new_agents = (
+                        len(self.exits[strategy.exit_class]) * entrance.prob
+                    )
+
+                for loc in self.pop.geography.locations.values():
+                    for race in self.params.classes.races:
+                        for i in range(
+                            round(
+                                num_new_agents
+                                * loc.ppl
+                                * loc.params.demographics[race].ppl
+                            )
+                        ):
+                            new_agent = self.pop.create_agent(loc, race, self.time)
+                            if entrance.age_in:
+                                new_agent.age = entrance.age
+                            self.pop.add_agent(new_agent)
+            elif entrance.enter_type == "replace":
+                assert (
+                    self.params.classes.exit[strategy.exit_class].exit_type != "none"
```

Comment:
while this is true, if the class is none, then we would end up with an empty list (assuming we switched the init to not exclude none) and then iterating over that would do nothing.  Is that cleaner? It feels more generic to me, which is nice, though you do lose the warning to an end user.

Reply:
I'm sort of mixed on it. I agree the cleaner is nice, so removing the error message is probably better.